### PR TITLE
[Kernel] Increase precision of GPTQ/AWQ Marlin kernel

### LINKED ISSUE #5793: [Bug]: Different quality responses using GPTQ / marlin kernels on A10 vs A100 GPUs
### 🐛 Describe the bug

Hello, 

I am running llama3-70b and mixtral with VLLM on a bunch of different kinds of machines. I encountered wildly different quality performance on A10 GPUs vs A100/H100 GPUs for ONLY gptq models and marlin kernels. GPTQ with marlin kernels is way faster than AWQ but with AWQ, i see roughly the same response on my test queries on either kind of GPU environment. For A10 deployments, the only difference in the settings is that I use 2 A10 24GB GPUs instead of 1 A100 or H100 (using the tensor parallelism param).  I am running vllm = 0.4.2. I am using examples from llama3-70b testing on a very simple test query but I also saw the similar flavor of quality issues with mixtral-awq vs mixtral-gptq as well and I also saw the same flavor of issues on all my other more complicated RAG test queries as well. Generally the gptq models on A10s are psychotic and unusable and behave as if the model files are screwed up, which is why i tested two publicly available quantizations of llama3... 

I also tried to simulate a GPU environment of 2 A10s on 1 single A100 by setting --gpu-memory-utilization to 0.55 i.e. I choked the vram on the A100 deployment to choke the kv-cache available space. I got the same good quality performance on the A100 that I always see albeit with less throughput since my kv-cache was choked. No psychotic responses like I see on A10s. 

Here's my examples,

Test prompt: "what model are you and what quantization do you use?"

Model : https://huggingface.co/casperhansen/llama-3-70b-instruct-awq
Model Args: --max-model-len 8192 --quantization awq --dtype half --gpu-memory-utilization 0.99 --trust-remote-code --enforce-eager --kv-cache-dtype fp8

Response on A10 or A100 (doesn't matter where I deploy it, I get pretty much the same response):
```
"I'm an AI assistant, and I'm happy to share some information about my architecture.

I'm a large language model, specifically a transformer-based model, which is a type of neural network designed primarily for natural language processing tasks. My model is based on the BERT (Bidirectional Encoder Representations from Transformers) architecture, which was developed by Google in 2018.

Regarding quantization, I don't use traditional quantization techniques like fixed-point or floating-point quantization. Instead, my model uses a technique called "knowledge distillation" to reduce the precision of my weights and activations while maintaining the accuracy of my responses.

Knowledge distillation is a method that involves training a smaller, simpler model (the "student") to mimic the behavior of a larger, more complex model (the "teacher"). In my case, the teacher model is a large, floating-point BERT model, and the student model is a smaller, integer-based version of myself.

During training, the student model learns to approximate the behavior of the teacher model by minimizing the difference between their outputs. This process allows me to maintain high accuracy while using fewer computational resources and reducing my memory footprint.

So, to summarize: I'm a BERT-based language model that uses knowledge distillation to reduce precision and improve efficiency, rather than traditional quantization techniques." 
```

Model : https://huggingface.co/TechxGenus/Meta-Llama-3-70B-GPTQ or https://huggingface.co/MaziyarPanahi/Meta-Llama-3-70B-Instruct-GPTQ (tested with both)
Model Args: --max-model-len 8192 --quantization marlin --dtype half --gpu-memory-utilization 0.99 --trust-remote-code --enforce-eager --kv-cache-dtype fp8

Response on A100:
```
"Thank you for asking!

I am a type of artificial intelligence language model, to be specific, I'm a transformer-based language model. My architecture is based on the transformer model introduced in the paper "Attention is All You Need" by Vaswani et al. in 2017.

As for quantization, I don't use traditional quantization methods like 8-bit or 16-bit integer quantization. Instead, I rely on a technique called "floating-point quantization" which allows me to represent my weights and activations using a combination of floating-point numbers and integers.

In more detail, my model uses a combination of the following techniques:

1. **Weight pruning**: This involves removing redundant or unnecessary weights from my neural network to reduce its size and computational requirements.
2. **Knowledge distillation**: This is a process where a smaller model (like me) is trained to mimic the behavior of a larger, more complex model. This helps to transfer knowledge from the larger model to me.
3. **Quantization-aware training**: During training, my weights and activations are simulated using lower precision data types (like integers) to mimic the effects of quantization. This helps me to learn how to represent information efficiently even with reduced precision.

These techniques allow me to maintain a good balance between accuracy and computational efficiency, I can provide helpful responses quickly while still being able to understand and respond accurately.

Please let me know if you have any other questions or if there's anything else I can help with!"
```

Response on 2 A10s (tensor parallelism):
```
"into the 2000s and the 2000s of the 2000s and the 2000s of the 2000s and the 2000s of the 2000s."
```
or something else equally psychotic and random (it produces a different response every time so I have a feeling this has to do with nondeterminism). Heres another example of a response on the same prompt:

```
"Anita; and the last; and the first; and the last; and the first; and the last.

Anita;
and the last;
and the first;
and the last;
and the first;
and the last.
[Anita];
[the last];
[the first];
[the last];
[the first];
[the last].
Anita;
and the last;
and the first;
and the last;
and the first;
and the last.
[Anita]];
[the last]];
[[the first]];
[[the last]];
[[the first]];
[[the last]].
Anita:;
and [[the last]];
[[the first]]];
[[the last]]];
[[the first]]];
[[the last]].
Anita:;
[[last]]];
[[first]]];
[[last]]];
[[first]];
[[last]].
Anita:;
 [[[last]]],
 [[[first]]],
 [[[last]]],
 [[[first]]],
 [[[last]]].
Anita:;
 [[[last]],
 [[[first]],
 [[[last]],
{{{first}},
{{{last}}].
Anita:;
{{{last}},
{{{first}},
{{{last}},
{{{first}},
{{{last}}].
Anita:;
{{{last}},
{{{first}},
{{{last}},
{{{first}},
{{{last}}].

This is a series of nested brackets,2019. The output is a series of nested brackets."
```

Let me know what you all think, I am very confused and surprised at this bug since both GPUs are ampere architecture and I wasn't able to pin the bug down on anything EXCEPT the difference in GPU in the deployment hardware. I've also been deploying and testing many different LLMs on a wide variety of hardware settings and have always been able to trust the heuristic that the quality of responses will be roughly the same but the speed of inference will be different.

### LINKED ISSUE #6258: [Bug]: tensor parallel (of 4 cards) gives bad answers in version 0.5.1 and later (compared to 0.4.1) with gptq marlin kernels (compared to gptq)
### Your current environment

sagemaker ml.g5.12xlarge instance (4 instances of a10g 24gb)
container is 
763104351884.dkr.ecr.us-west-2.amazonaws.com/djl-inference:0.27.0-deepspeed0.12.6-cu121
from
https://github.com/aws/deep-learning-containers/blob/master/available_images.md

### 🐛 Describe the bug

from vllm import LLM, SamplingParams
question = "what is the id of the team and what is the subtitute lineup of the home team for the match?"
history = str(["how many games the home team Sevilla won?"])
full_example = f"""
\n        \n\nYou are a transformation helper specialist, based on the history helping in transforming user input\nto a more structured and simpler text to a smaller model, which is less smart as you. \n\nMost of the times, the history could help you about entities which are now missing \nfrom the question\nTo illustrate the mission, if the user asked in the history about an entity (like \'Barcelona\'), and \nnow he asked about \'team\' (could be team, player, or other entity) or it seems to you that the an entity \nis missing in the context, perhaps the entity (\'Barcelona\') from the history could be the option to fill the gap. \n\nIf there is no entity in the history, please do not hallucinate and offer weird entity, for example if in the history\nyou saw \'home team\' and now he just mentioned \'team\', replace \'team\' with \'home team\' (applicable for away team too).\n\nWhen a replacement is occurred, please do not add \'the\' as part of the entity, just entity itself.\n\n        \nReturn it in valid JSON format according to the schema\n          \n\n User question: \n\n {question} \n\n History: \n\n {history}"""

prompts = [
    full_example
]
sampling_params = SamplingParams(temperature=0, top_p=0.95, max_new_tokens=256)

llm = LLM(model="TechxGenus/Meta-Llama-3-70B-Instruct-GPTQ", 'tokenizer_mode'='auto', 'gpu_memory_utilization'=0.7,  'guided_decoding_backend' ='lm-format-enforcer', tensor_parallel_size=4)

quantization not mentioned in order to get marlin kernels instead of standard gptq (if available).

outputs = llm.generate(prompts, sampling_params)

# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")

the output for version 0.4.1 is right
"what is the id of Sevilla and what is the substitute lineup of Sevilla for the match?"

for all versions forward (0.4.2 and above)
it returns weird answers (which does not make sense. like 1900 1900 1900 1900 and weird tokens in the answers)

Notes:
1.As a side check I ran everything the same, but with TechxGenus/Meta-Llama-3-8B-Instruct-AWQ, and it worked. As I suspect, when the shard is above the size of the gpu memory (24gb not enough for 70b gptq, compared to 8b awq), and a tp is needed, it does affect (maybe in v0.4.2 and forward you changed something in the code of the tensor parallel).

2. As another note, ran the same llama 70b gptq on 1 card v6000 48gb, everything worked great even with v0.5.1 (ofc 0.4.2 and up), so it almost 100% something with the megatron TP.

If another data needed, just comment and mention me.
Thanks

### PR DESCRIPTION
**Pull request by alexm-redhat**
FIX https://github.com/vllm-project/vllm/issues/5793 
FIX https://github.com/vllm-project/vllm/issues/6258

This PR increases the precision of Marlin kernel by modifying its "global_reduce" algorithm to use FP32 full-precision reduction, instead of the FP16 half-precision reduction that was used originally. We were able to implement the new FP32 global reduce efficiently, so that it introduces negligible overhead vs the original FP16 reduce.

The key idea is to introduce a temporary FP32 C buffer for the FP32 reduction (and not use the original FP16 C buffer as before). The new FP32 C buffer is limited in size based on the batch size (M dimension) and the potential "max_par" that can be achieved for each specific execution. Internally, each kernel thread-block detects on which thread-column-block it operates, and based on that accesses the appropriate chunk of the temporary C buffer in a fully thread-aligned way (to avoid any bank conflicts or non-contiguous memory reads/stores).

Here are micro-benchmark results for the gptq_marlin_gemm_fp16 vs gptq_marlin_gemm_fp32 (compared vs pytorch_gemm):

image

End-to-end performance verification on A100 shows max 5% penalty for 8b llama3 and no-penalty for 70b llama3.

### CODE REVIEW

**Comment by casper-hansen** [comment]
Do you have any data showcasing that this fixes the described accuracy issues? I'm guessing you could look at the examples referenced in the issue? And potentially run an evaluation of perplexity?

**Comment by alexm-neuralmagic** [comment]
@casper-hansen We had run internally a test on A10 for GSM dataset to verify accuracy, and we saw that with the old fp16 reduce the total accuracy is 58% and with the new fp32 reduce the accuracy is 73% (as it is supposed to be). Also, in the unit tests, we saw that the max difference is improved from e-3 to e-6 (almost double basically)

**Review by mgoin — APPROVED** [review]
Great work Alex, I'm glad it didn't require paring down split-k or any drastic changes. 

Do you think this should be added to fp8 marlin as well? We could also just wait for @LucasWilkinson 's type refactor

**Comment by alexm-neuralmagic** [comment]
I think it is not critical to add to fp8 marlin, since we did not had so much reports of accuracy issues like we had for GPTQ and especially AWQ (which is even more sensitive than GPTQ).
